docs/content/docs/deep/componets.md

```diff
@@ -0,0 +1,113 @@
+
+Currently no public APIs exist for defining new data-object types, but the existing experimental ones can be used in container schema.
+
+### Distributed Data Structures: DDSes
+
+A data structure in a container, with its own self contained editing and merge resolution logic.
+May contain handles to other DDSes.
+
+Some DDSes (mainly SHaredTree) provide their own functionality for further componentization of their content.
+
+While while the user of the container can choose which DDSes to load, DDSes themselves currently can not support being only partially loaded: either the whole DDS is available or not.
+The one exception to this is blobs which the user of the container can asynchronously upload then insert handle to them into the DDS.
+Future work (See SharedTree) may enable some DDSes to load subsets of their data on demand.
+
+## SharedTree
+
+SharedTree is design to encourage componentization by subtree / schema.
+Typical usage is a given component defines its schema and logic which works with that schema.
+This schema/component is then referenced/depended on by its parent, adding its schema as a child.
+
+If an application is using a design pattern with a separate Model and View,
+it should be practical to either package the view code for the component alongside its model or separately.
+
+If the same data (same schema) needs to be handled differently in different parts of the application, that logic (anything which differs) should be separate from the schema.
+Anything which is always the same however can be included as methods or additional state declared as part of the schema class if this is helpful.
+While the entire tre is branched, creating a branch can be done from just a subtree, so as long as the rest of the tree isn't modified, its indistinguishable from just branching the specific subtree.
+
+Tree provides its own branching feature, allowing additional local alternative views with buffered speculative edits.
+This does not however provide any way to branch other DDSes the tree references via handles, so everything which needs to be branched this way should be kept in the same tree.
+
+Undo and redo are at the SharedTree level, but since changes can be undone and redone in any order, the application should be able to have separate undo and redo stack for separate parts of the tree if desired.
+
+Shared Tree allows for subscribing to events for an entire subtree, as well on individual nodes, allowing each component to implement its own event handling.
+Due to how these events work, its recommenced to handle creation/removal/moving of components in the parent component which "owns" them (has them as a child).
+More generally, the modularity works best when a component just cares about the content of its subtree, and does not read or depend on its parent:
+access to the parent (Via `Tree.parent`) is supported, but correct invalidation code views using it is difficult and/or inefficient.
```

Comment:
What is the usage pattern for having this capability? This plays against composability / isolation

Reply:
Maybe other way to say it, usually there is a need for a child component to be able reach out to parent, but different interfaces should govern interactions between sub-components in the system.
I.e. in A -> B -> C dependency, B should be able to expose different interfaces to A and C.
The way I think about `parent` here (which might be incorrect) is that's the way to expose B to C (for C to reach to B). I think B should have direct control over this interaction.

The other question to ask (given references to move somewhere else): If A or B decides to move C around in the tree, I assume `C.parent` would change. Is there an event for that?

Reply:
Generally, customers, including our own Nick in our examples seems to really like parent access. If everyone adopted specific idealized design patterns that don't rely on it we wouldn't need it, but there are many perfectly reasonable and more intuitive (for some devs) patterns which rely on parent access.

In practice, people like using parent access for several things. For example some of our examples just provide the node in an array to the view for it, despite that view being able to delete or move it (ex: move to front), which is an edit on the parent. Thus the view needs to reach up to the parent to do those operations. That of course makes that view only work when it's tree is under a parent it understands. Instead passing the parent content into it behind some sort of abstraction which can handle the needed operations would be more of a pure module design, but that's annoying and in practice some devs haven't thought of that or don't want to bother. Thats why I'm pointing out you can avoid parent access for improved modularity here: it has dev time costs, but you can choose such patterns if your app needs them for modularity.

There are also use-cases, like looking up a node in an index, then checking where it is in the tree via its path.

Or holding onto a node for something like a selection or context for an undo, then later checking where it is to do something like scroll to it to make it visible.

Generally direct parent node access is one way for a child component to know about its context/owner. There are plenty of others which are more robust that apps can use (beyond just passing down the context explicitly), some of which build upon parent node access (ex: you can put a symbol on the parent node which the child can query to get a set of affordances about what that parent can do to the specific child in question. Then a node can ask things like: can my parent move me to front? If so, I'll add a button for that.).

> The other question to ask (given references to move somewhere else): If A or B decides to move C around in the tree, I assume C.parent would change. Is there an event for that?

Such a move would trigger a NodeChange event on B, since one of its fields is edited. Thus if C cares about its context (ex: read its parent) and wants to know if it moves, it needs to subscribe to an event on its parent for that. While this seems reasonable, it's not the most ergonomic or efficient way to find out if just which node is the parent changed. More targeted events for this specific use case are addressed in "Additional events" planned features below.

Reply:
Re "people like" angle: Some people like ability to have access to everything, but any big project eventually has rules in place dictating how parts of the app are written, including how to control dependencies / maintain flexibility (including ability to swap implementations).

I'd pose the following question: what do we expect people to be able to do with a parent like that?
From examples, I could conclude that we assume child needs to know the schema of the parent to use it. And if it knows the schema, that it has full control to parent capabilities - there is no way to limit / control it.